kpr: Support kube-apiserver HA

aditighag opened the pull request:
Support kube-apiserver high availability with kube-proxy replacement where the agent can fail over to an active kube-apiserver at runtime. 

### Background 
Cilium agent requires a connection to the kube-apiserver control plane to program the BPF datapath without depending on kube-proxy load-balancing when kube-proxy replacement is enabled. To achieve this, Cilium uses `API_SERVER_IP` and `API_SERVER_PORT` configurations for direct connection to the kube-apiserver. However, this approach doesn't support production environments that require multiple kube-apiservers for high availability. Additionally, it cannot rely on a fixed set of addresses provided at bootstrap time due to potential Kubernetes node failures or recycling in production environments.

### Commits summary
This PR introduces a flag for user to configure multiple kube-apiserver URLs. Additionally, it adds logic to connect to one of the active kube-apiservers in a cluster during initial connection time, and switches over to the `kubernetes` service at runtime (and post agent restarts).
The PR extends - https://github.com/cilium/cilium/pull/20090. 
See individual commits for implementation details.

![kubeapiserverHA](https://github.com/user-attachments/assets/847cc378-b588-44cd-9fb8-c46f4965c41a)

### Testing
- Manually tested on a kind cluster with >1 control plane nodes.
- Unit test cases cover various fail over scenarios. 

Follow-up:
Automated E2E test in a kind cluster environment.

Fixes: https://github.com/cilium/cilium/issues/19038

```release-note
Add support for kube-apiserver high availability with kube-proxy replacement where the Cilium agent can fail over to an active kube-apiserver at runtime.
```

aditighag committed:
pkg/k8s: Move rest config related code

This is a preparatory commit to move the code
involving rest configuration. Following commits
introduce logic that can mutate the configuration
at runtime to support high availability for
kube-apiservers. As a result, the rest configuration
management is encapsulated to a separate entity.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/k8s,service: Support kubeapi-server high availability with kpr

This commit introduces logic to support kubeapi-server high
availability with kube-proxy replacement by switching the
agent Kubernetes clients to connect to the `kubernetes`
service at runtime.

Cilium agent requires a connection to the kubeapi-server
control plane to program the BPF datapath without depending on
kube-proxy load-balancing when kube-proxy replacement is enabled.
To achieve this, Cilium uses API_SERVER_IP and API_SERVER_PORT
configurations for direct connection to the kubeapi-server.
However, this approach doesn't support production environments that
require multiple kubeapi-servers for high availability.
Additionally, it cannot rely on a fixed set of addresses provided
at bootstrap time due to potential Kubernetes node failures or recycling
in production environments.

Previous commits added the logic to connect to an active kube-apiserver
during bootstrap. Once the agent establishes an initial connection to
the API server, it receives runtime state for services including the
`kubernetes` service with API server endpoints. This state is written
to a file that's watched by the k8s-client module. It then validates
connectivity to the service virtual address, and updates the
internal HTTP round tripper to connect to the service address.
While kubelet exposes the API server virtual address in the
`KUBERNETES_SERVICE_HOST` address, this is overwritten when users set
`k8sServiceHost` config with kube-proxy replacement enabled.

The persisted state allows for the agent to restore the service and
endpoint addresses post agent restarts.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
install: Add flag to set multiple kube-apiservers

Users can set the flag to specify URLs of multiple
kube-apiserver instances in a cluster. Follow-up
commits implement the configuration so that the
agent Kubernetes clients can fail over to an active
instance at runtime.

Co-developed-by: Deepesh Pathak <deepeshpathak09@gmail.com>
Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/k8s, docs: Deprecate k8s-api-server flag

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/k8s: Rotate API server URL on failed heartbeats

The agent sends periodic heartbeats to the current
API server. Until the agent switches the kubernetes
client remote host to the `kubernetes` service address
for high availability, API servers need to be manually
rotated on failed heartbeats.
Once the switch over is done to the service address, manual
rotation isn't needed as the Cilium datapath will load balance
API requests to the active kube-api servers.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/k8s: Add benchmarks

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on pkg/k8s/watchers/service.go:
```diff
@@ -576,10 +577,28 @@ func (k *K8sServiceWatcher) addK8sSVCs(svcID k8s.ServiceID, oldSvc, svc *k8s.Ser
+		mapping.Endpoints[i] = svc.Backends[i].AddrString()
+	}
+	log.WithField("entry", mapping).Info("writing kubernetes service mapping")
+	_ = client.UpdateK8sAPIServerEntry(mapping)
```
@dylandreimerink Perhaps a better way for the watcher to update the k8s client cell about these k8s api service mappings is via stateDB (writing a new `statedb.RWTable[*tables.K8sServiceEndpointMapping]`)? Any comments?
Note that we'll still need to write these entries to a file so that we can restore them after agent restarts, but that logic can be encapsulated in `restConfigManager`.

dylandreimerink replied:
I don't think it makes much sense to create a table for just `K8sServiceEndpointMapping`, seeing as there is just 1 consumer and 1 producer.

You need to still make use of the service watcher to get the backends for a given server. However, with the experimental LB that is incoming, we should be able to let the `restConfigManager` directly watch the [services/frontends/backends tables](https://github.com/cilium/cilium/blob/main/pkg/loadbalancer/experimental/cell.go#L60).

It will still have to write to disk to persist it, but then its just one component, no communication over the FS anymore via file watchers.

aditighag replied:
Thanks! Let's go ahead with the current approach in the meantime as discussed.

aditighag committed:
pkg/k8s: Support setting multiple kubeapi-servers

This commit implements the logic to configure the
agent Kubernetes client module to connect to one
of the active kube-apiservers when users configure
multiple kube-apiserver URLs. This is a first step
to support high availability for the Kubernetes
API server. Follow-up commits switch over to the
`kubernetes` service with api server endpoints
at runtime.

The k8s-client cell provides a set of Kubernetes clients
for other modules to connect to the API server. More specifically,
the client provides Kubernetes client interfaces that are wired
into other modules during agent initialization via the hive/cell
framework. In order to fail over to an active kube-apiserver instance
at runtime, the commit wraps k8s client rest configuration with an
HTTP round tripper. The internal round tripper updates rest
configuration host whenever there are connectivity failures
to the api server.

Co-developed-by: Deepesh Pathak <deepeshpathak09@gmail.com>
Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>
Signed-off-by: Aditi Ghag <aditi@cilium.io>

thorn3r reviewed (approved):
Looks good for my codeowners :+1:

youngnick reviewed (approved):
Changes LGTM from sig-k8s.

dylandreimerink commented on pkg/k8s/client/restConfig_provider.go:
```diff
@@ -12,20 +15,43 @@ import (
+
+	// K8sAPIServerFilePath is the file path for storing kube-apiserver service and
+	// endpoints for high availability failover.
+	K8sAPIServerFilePath = "/run/cilium/state/k8sapi_server_state.json"
```
The path of `/run/cilium/state` is not set in stone. Would be better to use `option.Config.StateDir`

dylandreimerink commented on pkg/k8s/client/restConfig_provider.go:
```diff
@@ -215,3 +262,137 @@ func (r *restConfigManager) WrapRoundTripper(rt http.RoundTripper) http.RoundTri
+		case err := <-watcher.Errors:
+			health.Degraded(fmt.Sprintf("Failed to load  %q", K8sAPIServerFilePath), err)
+			r.log.WithError(err).Error("Unexpected error while watching kube-apiserver state file")
+			return err
```
So this means that if we get an error from the FS watcher, updates to the k8s API server are broken until the agent restarts.

A few things about this. First, the watcher can send and error, and still continue working. Only when the `watcher.Events` or `watcher.Errors` channel closes, has it actually stopped.

Second, might I suggest moving the creation of the FS watcher to inside of this job. And attempting to recreate the watcher if we error. With a timeout of course. That way, we have a chance to automatically recover if, for example, the file got deleted and then later re-create. 

Third, even if you want to keep this behavior, then you should still call `watcher.Close()`.

aditighag replied:
>might I suggest moving the creation of the FS watcher to inside of this job. And attempting to recreate the watcher if we error. 

That's a good suggestion. Done!

>First, the watcher can send and error, and still continue working. Only when the watcher.Events or watcher.Errors channel closes, has it actually stopped.

Gotcha! Updated the logic to not close the watcher on errors.

dylandreimerink commented on pkg/k8s/client/restConfig_provider.go:
```diff
@@ -215,3 +262,137 @@ func (r *restConfigManager) WrapRoundTripper(rt http.RoundTripper) http.RoundTri
+			r.apiServerURLs = updatedServerURLs
+		}
+	}
+}
```
nit: Why is this writting this way? Why do not a early return with `err != nil` and have the body of this if be the normal code path?

aditighag committed:
docs: Document kube-apiserver high availability support

Signed-off-by: Aditi Ghag <aditi@cilium.io>

squeed commented on pkg/k8s/client/restConfig_provider.go:
```diff
@@ -0,0 +1,425 @@
+		return
+	}
+
+	r.rt.Lock()
```
You have nested locks here; is there any chance of a deadlock? Is there anything else that takes `r.rt.Mutex` and `r.Mutex`?

aditighag replied:
No, as long as the locks are taken in the same order.

asauber reviewed (commented):
Have we considered that client-side load balancing for this feature might not be the right approach?

Kubernetes pods (including Cilium Agent) have the `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` environment variables, which allow for server-side load balancing of the Kubernetes API server.

This load balancing is then typically the responsibility of the Kubernetes control plane provider.

Example:
```
~ $ k exec -ti -n kube-system cilium-7eplp -c cilium-agent -- bash
root@ip-10-4-6-110:/home/cilium# echo "$KUBERNETES_SERVICE_HOST"
A93391BB28B5A90F2585A3E36A26C11F.sk1.us-east-2.eks.amazonaws.com
root@ip-10-4-6-110:/home/cilium# echo "$KUBERNETES_SERVICE_PORT"
443
```

asauber reviewed (approved):
CLI related changes look good
